mohanadarafe/GuideMe
# Outdoor direction implementation with Google Maps API

## Notice
Huge Feature implemented. The magnitude of this feature was criminally underestimated. We thought that Google API would take care of everything, we just needed to send the origin and destination. However, the whole UI and transitions and animations had to be built from scratch.
### Contents of the PR:
In this PR, quite a few features were implemented. First, using the Google Maps API, I was able to integrate outdoor directions in our application. This involves fetching route data from the mentioned API, and drawing it on the map while taking into consideration user preferences. Two main screens were created: a **PreviewDirections** screen, which shows a zoomed out itinerary on the map, the time and s start button on the bottom menu. By sliding up the bottom menu, the user can set their preferences, which will be taken into consideration only after the user presses the yellow refresh button on the Preview Direction Page. The next screen is the main **Directions** screen, which displays a step by step instruction of route between two _buildings_. or his location to another building. Also, the user can iterate through those instructions by using the right and left navigation arrows. Moreover, The **Directions** screens have a button allowing the user to zoom the map at his current location. Useful when the user wants to situate himself in the path.
### To summarize: 
**Two main screens were added:** : _PreviewDirections_ & _Directions_
**React Navigation was implemented for better handling of the screens.**
 **_Critical_ Bugs related to the layout and to the BottomMenu, DoubleSearch and Search components were fixed.**
**Secondary Screens** : _DoubleSearch_ & _PreferenceMenu_
Refactored components in _BottomMenu_ to be screens to take advantage of React Stack Navigation.
### Future Improvements and Current small issues
**To Fix:** 
-The header layout in the **Directions** screen is breaking when an instruction exceeds a certain length.
-The refresh button should only be clickable once after a selection was made in the Preference Menu.
- Bottom Menu in _Preview Directions_  gets out of bounds when the string of the time and distance is too long for its width.
- The transitions in the _Directions_ should center the map to the first point of a step instead of the whole polylines. 
- Disable _View Route_ in Double Search screen until the current location is fetched (async).
- Not all the files were thoroughly commented (for documentation purpose) as we focused on having those changes on master ASAP.

**Potential Addition**
- We have a lot of information returned by Google API concerning the transit details, maybe investigate what we could provide as extra for more detailed instructions.
 ### Not covered by this PR and this feature
- Google API doesn't allow routes for people with physical disabilities. May be allowed in the future. 
- Personas are also not covered and cannot be covered by this feature and PR.
- The shuttle bus is moved to the Side Menu as it doesn't interact with the Map.
- No Design Pattern implemented. Rather, React Pattern implemented (use of Hook and rendering techniques). We are too advanced in the project to redesign, the best alternative to have solid design patterns is to have them in mind when _**refactoring**_.
### How to test
Before starting, checkout the branch feature/outdoor-directions
`git checkout feature/outdoor-directions`
1. Start the application
`expo start` or `npm start`
2. Select a building in the Home Map Screen.
3. When Bottom Menu updates, click on the arrow up button and select "Get Directions".
4. Wait a few seconds before pressing View Route as the screen is fetching your current location.
5. You can change origin and destination.
6. Click on View Route.
7. You are now in the PreviewDirections screen and a zoomed out map containing the itinerary should be displayed
8. Slide up the bottom menu and select your preferences
9. Click on the down arrow
10. Click on the "Start" button
11. You are now in the Directions screen
12. It should display the starting position (address) and the destination.
13. Click the right arrow (bottom right corner) to navigate through the directions.
14. The map should be updated with every click.
## GIFS
#### Preview Direction and Detailed Directions
![Animated GIF-downsized_large](https://user-images.githubusercontent.com/33582344/77288490-571f4d80-6cae-11ea-9b79-548bdc410dd4.gif)

#### Preference Menu and Update Direction
![Animated GIF-downsized_large (1)](https://user-images.githubusercontent.com/33582344/77288579-846bfb80-6cae-11ea-90ac-5723bf09577c.gif)

## Mentions
I also want to give honorable mentions to @csbduzi & @bkenza  as they actively coauthored this PR as you would be able to see in the commits. This PR wouldn't have come in time without them. THANK YOU A LOT <3
